src/snowflake/snowpark/mock/_udf.py

```diff
@@ -10,21 +11,71 @@
 )
 from snowflake.snowpark._internal.utils import TempObjectType
 from snowflake.snowpark.exceptions import SnowparkSQLException
+from snowflake.snowpark.mock._stage_registry import extract_stage_name_and_prefix
 from snowflake.snowpark.types import DataType
 from snowflake.snowpark.udf import UDFRegistration, UserDefinedFunction
 
 
 class MockUDFRegistration(UDFRegistration):
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
-        self._registry: Dict[str, Callable] = dict()
+        self._registry: Dict[
+            str, Union[Callable, Tuple[str, str]]
+        ] = (
+            dict()
+        )  # maps udf name to either the callable or a pair of str (module_name, callable_name)
+        self._udf_level_imports = dict()
+        self._session_level_imports = set()
 
-    def register_from_file(self, *_, **__) -> UserDefinedFunction:
-        self._session._conn.log_not_supported_error(
-            external_feature_name="udf",
-            error_message="Registering UDF from file is not currently supported.",
-            raise_error=NotImplementedError,
-        )
+    def _clear_session_imports(self):
+        self._session_level_imports.clear()
+
+    def _import_python_file(
+        self,
+        file_path: str,
+        import_path: Optional[str] = None,
+        udf_name: Optional[str] = None,
+    ) -> str:
+        """
+        Imports a python file or a directory of python module structure or a zip of the former.
+        Returns the name of the Python module to be imported.
+        When udf_name is not None, the import is added to the UDF associated with the name;
+        Otherwise, it is a session level import and will be added to every future UDF call.
+        """
+        file_name, file_extension = os.path.splitext(os.path.basename(file_path))
+
+        if file_path.startswith("@"):  # file is on a stage
+            stage_registry = self._session._conn.stage_registry
```

Comment:
maybe add `snow://` too here:

https://github.com/snowflakedb/snowpark-python/pull/1346

Reply:
I created https://snowflakecomputing.atlassian.net/browse/SNOW-1285387 to track this